src/Microsoft.AspNetCore.Server.IISIntegration/NativeMethods.cs

```diff
@@ -37,85 +29,212 @@ public enum REQUEST_NOTIFICATION_STATUS
+            return iisConfigurationData;
+        }
+
+        public static bool HttpTryGetServerVariable(IntPtr pInProcessHandler, string variableName, out string value)
+        {
+            return http_get_server_variable(pInProcessHandler, variableName, out value) == 0;
+        }
+
+        public static bool HttpShutdown()
+        {
+            return http_shutdown();
+        }
+
+        public static unsafe int HttpWebsocketsReadBytes(IntPtr pInProcessHandler, byte* pvBuffer, int cbBuffer, PFN_WEBSOCKET_ASYNC_COMPLETION pfnCompletionCallback, IntPtr pvCompletionContext, out int dwBytesReceived, out bool fCompletionExpected)
+        {
+            return http_websockets_read_bytes(pInProcessHandler, pvBuffer, cbBuffer, pfnCompletionCallback, pvCompletionContext, out dwBytesReceived, out fCompletionExpected);
+        }
+
+        public static unsafe int HttpWebsocketsWriteBytes(IntPtr pInProcessHandler, HttpApiTypes.HTTP_DATA_CHUNK* pDataChunks, int nChunks, PFN_WEBSOCKET_ASYNC_COMPLETION pfnCompletionCallback, IntPtr pvCompletionContext, out bool fCompletionExpected)
+        {
+            return http_websockets_write_bytes(pInProcessHandler, pDataChunks, nChunks, pfnCompletionCallback, pvCompletionContext, out fCompletionExpected);
+        }
+
+        public static void HttpEnableWebsockets(IntPtr pInProcessHandler)
+        {
+            Validate(http_enable_websockets(pInProcessHandler));
+        }
+
+        public static void HttpCancelIO(IntPtr pInProcessHandler)
+        {
+            Validate(http_cancel_io(pInProcessHandler));
+        }
+
+        public static unsafe void HttpResponseSetUnknownHeader(IntPtr pInProcessHandler, byte* pszHeaderName, byte* pszHeaderValue, ushort usHeaderValueLength, bool fReplace)
+        {
+            Validate(http_response_set_unknown_header(pInProcessHandler, pszHeaderName, pszHeaderValue, usHeaderValueLength, fReplace));
+        }
+
+        public static unsafe void HttpResponseSetKnownHeader(IntPtr pInProcessHandler, int headerId, byte* pHeaderValue, ushort length, bool fReplace)
+        {
+            Validate(http_response_set_known_header(pInProcessHandler, headerId, pHeaderValue, length, fReplace));
+        }
+
+        public static void HttpGetAuthenticationInformation(IntPtr pInProcessHandler, out string authType, out IntPtr token)
+        {
+            Validate(http_get_authentication_information(pInProcessHandler, out authType, out token));
+        }
 
         [DllImport("kernel32.dll")]
-        public static extern IntPtr GetModuleHandle(string lpModuleName);
+        private static extern IntPtr GetModuleHandle(string lpModuleName);
 
         public static bool is_ancm_loaded()
         {
             return GetModuleHandle(AspNetCoreModuleDll) != IntPtr.Zero;
         }
+
+        private static void Validate(int hr)
+        {
+            if (hr != 0)
+            {
+                throw Marshal.GetExceptionForHR(hr);
```

Comment:
We should wrap these exceptions somewhere with a custom exception because most of these will be ugly COM exceptions.

Reply:
We can't really add more information to it, HRESULT and stack trace is all we have.